src/urllib3/contrib/pyopenssl.py

```diff
@@ -366,9 +366,11 @@ def sendall(self, data: bytes) -> None:
             )
             total_sent += sent
 
-    def shutdown(self) -> None:
-        # FIXME rethrow compatible exceptions should we ever use this
-        self.connection.shutdown()
+    def shutdown(self, how: int) -> None:
+        try:
+            self.connection.shutdown(how)
```

Comment:
It looks like this line is not reached in tests because pyOpenSSL doesn't accept any parameters for `shutdown`
I tried changing `HTTPConnectionPool` to `HTTPSConnectionPool` in the new test, but it just hanged, not sure why

Reply:
<details>
<summary>Ah, it didn't set up SSL use in the socket handler</summary>

```diff
diff --git a/test/with_dummyserver/test_socketlevel.py b/test/with_dummyserver/test_socketlevel.py
index 46feb4bf..5ffbb56c 100644
--- a/test/with_dummyserver/test_socketlevel.py
+++ b/test/with_dummyserver/test_socketlevel.py
@@ -94,7 +94,7 @@ class TestSNI(SocketDummyServerTestCase):
             sock.close()
 
         self._start_server(socket_handler)
-        with HTTPSConnectionPool(self.host, self.port) as pool:
+        with HTTPSConnectionPool(self.host, self.port, ca_certs=DEFAULT_CA) as pool:
             try:
                 pool.request("GET", "/", retries=0)
             except MaxRetryError:  # We are violating the protocol
@@ -1019,14 +1019,24 @@ class TestSocketClosing(SocketDummyServerTestCase):
 
         def socket_handler(listener: socket.socket) -> None:
             sock = listener.accept()[0]
+            try:
+                ssl_sock = original_ssl_wrap_socket(
+                    sock,
+                    server_side=True,
+                    keyfile=DEFAULT_CERTS["keyfile"],
+                    certfile=DEFAULT_CERTS["certfile"],
+                    ca_certs=DEFAULT_CA,
+                )
+            except (ssl.SSLError, OSError):
+                return
 
             # Consume request
             buf = b""
             while not buf.endswith(b"\r\n\r\n"):
-                buf = sock.recv(65535)
+                buf = ssl_sock.recv(65535)
 
             # Send incomplete message (note Content-Length)
-            sock.send(
+            ssl_sock.send(
                 b"HTTP/1.1 200 OK\r\n"
                 b"Content-Type: text/plain\r\n"
                 b"Content-Length: 10\r\n"
@@ -1034,11 +1044,12 @@ class TestSocketClosing(SocketDummyServerTestCase):
                 b"Hi-"
             )
             timed_out.wait(5)
+            ssl_sock.close()
             sock.close()
 
         self._start_server(socket_handler)
 
-        with HTTPConnectionPool(self.host, self.port) as pool:
+        with HTTPSConnectionPool(self.host, self.port, ca_certs=DEFAULT_CA) as pool:
             response = pool.urlopen("GET", "/", preload_content=False, retries=0)
             # Calling shutdown here calls shutdown() on the underlying socket,
             # so that the remaining read will fail instead of blocking
```
</details>

```suggestion
            self.connection.shutdown()
```

Reply:
Should we test this with and without SSL?

Reply:
I had missed this comment, sorry. We should test with SSL, but I'd like to know if there's value in testing without SSL. I'll adapt my new test to use SSL tomorrow.

Reply:
OK, so that was 4 days ago, sorry, but it's now done in [`f6eab0b` (#3527)](https://github.com/urllib3/urllib3/pull/3527/commits/f6eab0b67e7c3ccc7250e2a22014922596c11f23). Please take another look.